File: Assets/Scripts/Pet/PetNeedsSystem.cs
using UnityEngine;

namespace MyPet.Pet
{
    public enum PetLifeStage
    {
        Baby,
        Adult
    }

    [System.Serializable]
    public class PetNeedsSnapshot
    {
        public int ageDays;
        public PetLifeStage lifeStage;
        public float hunger;
        public float energy;
        public float mood;
        public float bond;
        public float socialNeed;
    }

    /// <summary>
    /// 宠物需求与成长模型（MVP 版）
    /// 0 = 极低，100 = 极高。
    /// </summary>
    public class PetNeedsSystem : MonoBehaviour
    {
        [Header("Identity")]
        [SerializeField] private string petId = "pet_001";

        [Header("Growth")]
        [SerializeField] private int ageDays = 1;
        [SerializeField] private PetLifeStage lifeStage = PetLifeStage.Baby;
        [SerializeField] private float growthSecondsPerDay = 120f;

        [Header("Needs (0~100)")]
        [SerializeField] private float hunger = 25f;
        [SerializeField] private float energy = 80f;
        [SerializeField] private float mood = 70f;
        [SerializeField] private float bond = 40f;
        [SerializeField] private float socialNeed = 30f;

        [Header("Decay Rates / sec")]
        [SerializeField] private float hungerIncreasePerSec = 0.35f;
        [SerializeField] private float energyDecreasePerSec = 0.2f;
        [SerializeField] private float moodDecreasePerSec = 0.08f;
        [SerializeField] private float socialNeedIncreasePerSec = 0.15f;

        private float growthTimer;

        public string PetId => petId;
        public int AgeDays => ageDays;
        public PetLifeStage LifeStage => lifeStage;
        public float Hunger => hunger;
        public float Energy => energy;
        public float Mood => mood;
        public float Bond => bond;
        public float SocialNeed => socialNeed;

        private void Update()
        {
            TickNeeds(Time.deltaTime);
            TickGrowth(Time.deltaTime);
        }

        public void TickNeeds(float dt)
        {
            hunger = Clamp01To100(hunger + hungerIncreasePerSec * dt);
            energy = Clamp01To100(energy - energyDecreasePerSec * dt);
            mood = Clamp01To100(mood - moodDecreasePerSec * dt);
            socialNeed = Clamp01To100(socialNeed + socialNeedIncreasePerSec * dt);

            if (hunger > 80f || energy < 20f)
            {
                mood = Clamp01To100(mood - 0.15f * dt);
            }
        }

        public void TickGrowth(float dt)
        {
            growthTimer += dt;
            if (growthTimer < growthSecondsPerDay)
            {
                return;
            }

            growthTimer = 0f;
            ageDays += 1;
            lifeStage = ageDays >= 7 ? PetLifeStage.Adult : PetLifeStage.Baby;
        }

        public void Feed(float amount = 30f)
        {
            hunger = Clamp01To100(hunger - amount);
            mood = Clamp01To100(mood + 8f);
            bond = Clamp01To100(bond + 2f);
        }

        public void Petting(float amount = 10f)
        {
            mood = Clamp01To100(mood + amount);
            bond = Clamp01To100(bond + 3f);
            socialNeed = Clamp01To100(socialNeed - 6f);
        }

        public void Play(float amount = 12f)
        {
            mood = Clamp01To100(mood + amount);
            energy = Clamp01To100(energy - 8f);
            socialNeed = Clamp01To100(socialNeed - 10f);
            bond = Clamp01To100(bond + 4f);
        }

        public void Sleep(float amount = 20f)
        {
            energy = Clamp01To100(energy + amount);
            mood = Clamp01To100(mood + 2f);
        }

        public PetNeedsSnapshot GetSnapshot()
        {
            return new PetNeedsSnapshot
            {
                ageDays = ageDays,
                lifeStage = lifeStage,
                hunger = hunger,
                energy = energy,
                mood = mood,
                bond = bond,
                socialNeed = socialNeed
            };
        }

        public void ApplySnapshot(PetNeedsSnapshot snapshot)
        {
            ageDays = snapshot.ageDays;
            lifeStage = snapshot.lifeStage;
            hunger = Clamp01To100(snapshot.hunger);
            energy = Clamp01To100(snapshot.energy);
            mood = Clamp01To100(snapshot.mood);
            bond = Clamp01To100(snapshot.bond);
            socialNeed = Clamp01To100(snapshot.socialNeed);
        }

        private static float Clamp01To100(float value)
        {
            return Mathf.Clamp(value, 0f, 100f);
        }
    }
}

File: src/core/PetNeedsSystem.js
export class PetNeedsSystem {
  constructor() {
    this.petId = 'pet_001';
    this.ageDays = 1;
    this.lifeStage = 'baby';

    this.hunger = 25;
    this.energy = 80;
    this.mood = 70;
    this.bond = 40;
    this.socialNeed = 30;

    this._growthTimer = 0;
    this._growthSecondsPerDay = 120;
  }

  update(dt) {
    this.hunger = clamp(this.hunger + 0.35 * dt);
    this.energy = clamp(this.energy - 0.2 * dt);
    this.mood = clamp(this.mood - 0.08 * dt);
    this.socialNeed = clamp(this.socialNeed + 0.15 * dt);

    if (this.hunger > 80 || this.energy < 20) {
      this.mood = clamp(this.mood - 0.15 * dt);
    }

    this._growthTimer += dt;
    if (this._growthTimer >= this._growthSecondsPerDay) {
      this._growthTimer = 0;
      this.ageDays += 1;
      this.lifeStage = this.ageDays >= 7 ? 'adult' : 'baby';
    }
  }

  feed(amount = 30) {
    this.hunger = clamp(this.hunger - amount);
    this.mood = clamp(this.mood + 8);
    this.bond = clamp(this.bond + 2);
  }

  pet(amount = 10) {
    this.mood = clamp(this.mood + amount);
    this.bond = clamp(this.bond + 3);
    this.socialNeed = clamp(this.socialNeed - 6);
  }

  play(amount = 12) {
    this.mood = clamp(this.mood + amount);
    this.energy = clamp(this.energy - 8);
    this.socialNeed = clamp(this.socialNeed - 10);
    this.bond = clamp(this.bond + 4);
  }

  sleep(amount = 20) {
    this.energy = clamp(this.energy + amount);
    this.mood = clamp(this.mood + 2);
  }

  getSnapshot() {
    return {
      petId: this.petId,
      ageDays: this.ageDays,
      lifeStage: this.lifeStage,
      hunger: this.hunger,
      energy: this.energy,
      mood: this.mood,
      bond: this.bond,
      socialNeed: this.socialNeed,
    };
  }

  applySnapshot(snapshot) {
    this.petId = snapshot.petId ?? this.petId;
    this.ageDays = snapshot.ageDays ?? this.ageDays;
    this.lifeStage = snapshot.lifeStage ?? this.lifeStage;
    this.hunger = clamp(snapshot.hunger ?? this.hunger);
    this.energy = clamp(snapshot.energy ?? this.energy);
    this.mood = clamp(snapshot.mood ?? this.mood);
    this.bond = clamp(snapshot.bond ?? this.bond);
    this.socialNeed = clamp(snapshot.socialNeed ?? this.socialNeed);
  }
}

function clamp(v) {
  return Math.max(0, Math.min(100, v));
}
